airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/MemoryManager.kt

```diff
@@ -17,19 +17,19 @@ import kotlin.concurrent.withLock
  * TODO: Some degree of logging/monitoring around how accurate we're actually being?
  */
 @Singleton
-class MemoryManager {
-    private val availableMemoryBytes: Long = Runtime.getRuntime().maxMemory()
+class MemoryManager(availableMemoryProvider: AvailableMemoryProvider) {
+    private val availableMemoryBytes: Long = availableMemoryProvider.availableMemoryBytes
     private var usedMemoryBytes = AtomicLong(0L)
-    private val memoryLock = ReentrantLock()
-    private val memoryLockCondition = memoryLock.newCondition()
+    private val releaseUpdates = Channel<Unit>(Channel.UNLIMITED)
+
+    val remainingMemoryBytes: Long
+        get() = availableMemoryBytes - usedMemoryBytes.get()
 
     suspend fun reserveBlocking(memoryBytes: Long) {
-        memoryLock.withLock {
-            while (usedMemoryBytes.get() + memoryBytes > availableMemoryBytes) {
-                memoryLockCondition.await()
-            }
-            usedMemoryBytes.addAndGet(memoryBytes)
+        while (usedMemoryBytes.get() + memoryBytes > availableMemoryBytes) {
+            releaseUpdates.receive()
```

Comment:
if `memoryBytes > availableMemoryBytes`, we just wait forever, no?

Reply:
this isn't memory safe. 2 threads could allocate more memory than is available.

Reply:
also, as soon as I typed `if`, copilot knew what I was doing :)